azkaban-common/src/main/java/azkaban/executor/container/ContainerizedExecutionManager.java

```diff
@@ -0,0 +1,383 @@
+    return DispatchMethod.CONTAINERIZED;
+  }
+
+  @Override
+  public void start() {
+    this.queueProcessor = setupQueueProcessor();
+    this.queueProcessor.start();
+  }
+
+  private QueueProcessorThread setupQueueProcessor() {
+    return new QueueProcessorThread(
+        this.azkProps, this.executorLoader);
+  }
+
+  @Override
+  public Status getStartStatus() {
+    return Status.READY;
+  }
+
+  @Override
+  public void shutdown() {
+    if (null != this.queueProcessor) {
+      this.queueProcessor.shutdown();
+    }
+  }
+
+  @Override
+  public void enableQueueProcessorThread() {
+    this.queueProcessor.setActive(true);
+  }
+
+  @Override
+  public void disableQueueProcessorThread() {
+    this.queueProcessor.setActive(false);
+  }
+
+  public State getQueueProcessorThreadState() {
+    return this.queueProcessor.getState();
+  }
+
+  public boolean isQueueProcessorThreadActive() {
+    return this.queueProcessor.isActive();
+  }
+
+  /**
+   * This QueueProcessorThread will fetch executions from database in batch and dispatch it in
+   * container.
+   */
+  private class QueueProcessorThread extends Thread {
+
+    private final long queueProcessorWaitInMS;
+    private final Props azkProps;
+    private volatile boolean shutdown = false;
+    private volatile boolean isActive = true;
+    private ExecutorService executorService;
+    private ExecutorLoader executorLoader;
+    private boolean executionsBatchProcessingEnabled;
+    private int executionsBatchSize;
+
+    public QueueProcessorThread(final Props azkProps, final ExecutorLoader executorLoader) {
+      this.azkProps = azkProps;
+      this.executorLoader = executorLoader;
+      setActive(
+          this.azkProps.getBoolean(Constants.ConfigurationKeys.QUEUEPROCESSING_ENABLED, true));
+      this.queueProcessorWaitInMS =
+          this.azkProps.getLong(Constants.ConfigurationKeys.QUEUE_PROCESSOR_WAIT_IN_MS, 1000);
+      this.executionsBatchProcessingEnabled = azkProps
+          .getBoolean(ContainerizedExecutionManagerProperties.CONTAINERIZED_EXECUTION_BATCH_ENABLED,
+              false);
+      this.executionsBatchSize =
+          azkProps
+              .getInt(ContainerizedExecutionManagerProperties.CONTAINERIZED_EXECUTION_BATCH_SIZE,
+                  10);
+
+      this.executorService = Executors.newFixedThreadPool(azkProps.getInt(
+          ContainerizedExecutionManagerProperties.CONTAINERIZED_EXECUTION_PROCESSING_THREAD_SIZE,
+          10));
+      this.setName("Containerized-QueueProcessor-Thread");
+    }
+
+    @Override
+    public void run() {
+      // Loops till QueueProcessorThread is shutdown
+      while (!this.shutdown) {
+        synchronized (this) {
+          try {
+            // start processing queue if active, otherwise wait for sometime
+            if (this.isActive) {
+              processQueuedFlows();
+            }
+            wait(queueProcessorWaitInMS);
+          } catch (final Exception e) {
```

Comment:
If I understand correctly, the intention is to provide some form of rate-limiting when multiple flows are waiting to be dispatched. 
An alternative to dispatching a fixed number of flows followed by wait could be to use a model that only guarantees that 'no more than `x` flows will be dispatched within a time interval `t`'.
This is already available as part of guava: https://guava.dev/releases/19.0/api/docs/index.html?com/google/common/util/concurrent/RateLimiter.html

Please see if it simplifies this code and (perhaps) better reflects the  requirement.

Reply:
Great suggestion! Thanks.